asynctest/mock.py

```diff
@@ -383,12 +383,62 @@ class MagicMock(AsyncMagicMixin, unittest.mock.MagicMock,
     """
 
 
-class _AwaitEvent(asyncio.Event):
-    """
-    A mix between asyncio.Event and bool.
-    """
+class _AwaitEvent:
+    def __init__(self, mock):
+        self._mock = mock
+        self._condition = asyncio.Condition()
+
+    @asyncio.coroutine
+    def wait(self, skip=0):
+        """
+        Wait for await to be called at least ``skip`` number of times.
+        """
+        def predicate(mock):
+            return mock.await_count > skip
+
+        return (yield from self.wait_for(predicate))
+
+    @asyncio.coroutine
+    def wait_next(self, skip=0):
+        """
+        Wait for await to be called at least ``skip + 1`` more times.
+        """
+        await_count = self._mock.await_count
+
+        def predicate(mock):
+            return mock.await_count > await_count + skip
+
+        return (yield from self.wait_for(predicate))
+
+    @asyncio.coroutine
+    def wait_for(self, predicate):
```

Comment:
@Martiusweb I think github's parser ate the prefix, what did you mean `_`?

Reply:
Interresting use case. I didn't think of something like this.

I think that we can add this example in the documentation, and tell that the predicate is only checked when the mock is awaited. I don't want users to have the feeling that it can magically awake after a random event.

Reply:
In fact, I was thinking about this:

    def predicate(mock):
        return mock.await_count > 0 and m.call_args == (("foo", ), {})

    t1 = asyncio.ensure_future(mock("foo"))
    t2 = mock("bar")  # not yet scheduled
    mock.awaited.wait_for(predicate)  # will hang forever: mock.call_args is now based on the last call.

The problem is that the predicate is based on the mock (ie coroutine function), not the coroutine that will be awaited itself.

I'm affraid that this will be confusing as it's not possible to wait_for() a given instance of the coroutine function to be awaited. Maybe we should keep this private and let users write their own logic when they want to do complex things (one can use the ``wraps`` argument to customize th behavior of the mock).

Reply:
You should use `call_args_list`.

Reply:
It doesn't solve the issue of knowing which instance you're awaiting:

    def predicate_wait_for_bar(mock):
        return mock.await_count > 0 and (("bar", ), {}) in m.call_args_list

    t1 = asyncio.ensure_future(mock("foo"))
    t2 = mock("bar")  # not yet scheduled
    mock.awaited.wait_for(predicate_wait_for_bar)  # predicate will be true

Reply:
You should find an index of the call in `call_args_list` and compare it against `await_count`.

Reply:
This won’t be 100% accurate (you can await in different order) but could be sufficient for some tests.

Reply:
In other hand for those tests `skip` would be sufficient as well. Perhaps we should add `await_args` / `await_args_list`?

Reply:
I think that the use case for checking which exact call has been awaited makes sense, but I'm not conviced with generic/public wait_for() method yet.

I think that we should merge this PR with a private _wait_for() as it's already very useful like this. Then we can work on a new PR to implement wait_call() and wait_next_call() or add optional parameters matching in wait() and wait_next().

Reply:
When I was working on that, ability to await a specific call was only one of use cases I thought of. Another was an ability to await an external consition like setting mock’s attribute to a specific value. Or more general: any modification of mock’s state related to await. I also thought of ability to await external conditions, but couldn’t think of a practical example.

I would opt for having `wait_for` public as it seems in line with the naming of async primitives. People who would confuse it would probably have problems with `wait` and `wait_next` as well. This an advanced method, but it is possible to mark it as such with docs.
 
What do you say if we keep it public via naming, but exclude from the documentation for now?

Reply:
Please check https://github.com/GreatFruitOmsk/asynctest/commit/3b22978c031ee23a222546ff3e36538a0cb3edaf, it introduces an independent list of "await args" and among other address the issue in your example for `wait_for`.